user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserQuotasOperator.java

```diff
@@ -202,119 +92,105 @@ boolean isPathExist(String path)    {
+        if (current != null) {
             log.debug("Deleting quotas for user {}", username);
-            JsonObject deleteJson = removeQuotasFromJsonUser(data);
-            if (configJsonIsEmpty(deleteJson)) {
-                zkClient.deleteRecursive("/config/users/" + encodedUsername);
-                log.debug("User {} deleted from ZK store", username);
-            } else {
-                zkClient.writeData("/config/users/" + encodedUsername, deleteJson.toBuffer().getBytes());
-            }
-            notifyChanges(username);
+            current.setProducerByteRate(null);
+            current.setConsumerByteRate(null);
+            current.setRequestPercentage(null);
+            alterUserQuotas(username, toClientQuotaAlterationOps(current));
         } else {
             log.warn("Quotas for user {} already don't exist", username);
         }
     }
 
-    /**
-     * Deletes the quotas from existing JSON
-     *
-     * @param userConfig JSON string with existing userConfig configuration as byte[]
-     *
-     * @return  Returns the updated JSON without the quotas
-     */
-    protected JsonObject removeQuotasFromJsonUser(byte[] userConfig)   {
-        JsonObject json = new JsonObject(new String(userConfig, StandardCharsets.UTF_8));
+    protected void alterUserQuotas(String username, Set<ClientQuotaAlteration.Op> ops) throws Exception {
+        ClientQuotaEntity cqe = new ClientQuotaEntity(Collections.singletonMap(ClientQuotaEntity.USER, username));
+        ClientQuotaAlteration cqa = new ClientQuotaAlteration(cqe, ops);
+        try {
+            adminClient.alterClientQuotas(Collections.singleton(cqa)).all().get();
+        } catch (Exception e) {
+            log.error("Creating/Altering quotas for user {} failed", username, e);
+            throw e;
+        }
+    }
 
-        validateJsonVersion(json);
-        JsonObject config = json.getJsonObject("config");
-        if (config == null) {
-            json.put("config", new JsonObject());
-        } else {
-            if (config.getString("producer_byte_rate") != null) {
-                config.remove("producer_byte_rate");
-            }
-            if (config.getString("consumer_byte_rate") != null) {
-                config.remove("consumer_byte_rate");
-            }
-            if (config.getString("request_percentage") != null) {
-                config.remove("request_percentage");
+    protected KafkaUserQuotas describeUserQuotas(String username) throws Exception {
+        ClientQuotaFilterComponent c = ClientQuotaFilterComponent.ofEntity(ClientQuotaEntity.USER, username);
+        ClientQuotaFilter f =  ClientQuotaFilter.contains(Collections.singleton(c));
+        KafkaUserQuotas current = null;
+        try {
+            ClientQuotaEntity cqe = new ClientQuotaEntity(Collections.singletonMap(ClientQuotaEntity.USER, username));
+            Map<ClientQuotaEntity, Map<String, Double>> map = adminClient.describeClientQuotas(f).entities().get();
```

Comment:
Should this be done asynchronously? Or is this whole block in some `executeBlocking`?

Reply:
Yes, all the adminClient calls are done from inside the `reconcile` method (which calls `createOrUpdate`) and the `reconcile` does these calls in an `executeBlocking`. The same happens for the `SimpleAclOperator` for example.